api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,213 @@
+	//
+	// If clusterSelector is also set, then the selector as a whole selects
+	// Machines matching selector belonging to Clusters selected by clusterSelector.
+	// If clusterSelector is not set, it selects all Machines matching selector in
+	// all Clusters.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// clusterSelector is a label selector which selects Machines by the labels of
+	// their Clusters.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects Machines of all Clusters.
+	//
+	// If selector is also set, then the selector as a whole selects
+	// Machines matching selector belonging to Clusters selected by clusterSelector.
+	// If selector is not set, it selects all Machines belonging to Clusters
+	// selected by clusterSelector.
+	// +optional
+	ClusterSelector *metav1.LabelSelector `json:"clusterSelector,omitempty"`
+}
+
+type MachineDrainRulePodSelector struct { //nolint:revive // Intentionally not adding a godoc comment as it would show up additionally to the field comment in the CRD
+	// selector is a label selector which selects Pods by their labels.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects all Pods.
+	//
+	// If namespaceSelector is also set, then the selector as a whole selects
+	// Pods matching selector in Namespaces selected by namespaceSelector.
+	// If namespaceSelector is not set, it selects all Pods matching selector in
+	// all Namespaces.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// namespaceSelector is a label selector which selects Pods by the labels of
+	// their Namespaces.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects Pods of all Namespaces.
+	//
+	// If selector is also set, then the selector as a whole selects
+	// Pods matching selector in Namespaces selected by namespaceSelector.
+	// If selector is not set, it selects all Pods in Namespaces selected by
+	// namespaceSelector.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machinedrainrules,shortName=mdr,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+// +kubebuilder:printcolumn:name="Behavior",type="string",JSONPath=".spec.drain.behavior",description="Drain behavior"
+// +kubebuilder:printcolumn:name="Order",type="string",JSONPath=".spec.drain.order",description="Drain order"
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp",description="Time duration since creation of the MachineDrainRule"
+
+// MachineDrainRule is the Schema for the MachineDrainRule API.
+type MachineDrainRule struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec MachineDrainRuleSpec `json:"spec,omitempty"`
```

Comment:
Missing a required tag? And shouldn't have omitempty?

Reply:
Every single CRD that we have is using exactly thish for spec and status. Not sure if we should start making this inconsistent now

Reply:
Seems like 71 vs 30 in k8s.io/api are using `json:spec,omitempty` (not sure if my grep is correct). What am I missing? :)

Reply:
What is the point in a CRD that looks like
```
apiVersion: blah
kind: baz
metadata:
  name: foo
```
What does the controller do if there's no spec?

IMO, spec should always be required, especially when it has required fields within. Else, you've got to follow the optional guidance, which means it should be a pointer, which would lead to other inconsistencies

> Not sure if we should start making this inconsistent now

I love being consistent, but please don't let consistency stand in the way of progress towards better APIs, and following conventions that have evolved over time. Lets not do something and repeat a past mistake, just to be consistent.

Reply:
Done. I assume same for the List?

Reply:
Same for list.

For objectmeta, I've never marked it required, but you're right, realistically it should be.

I think the API server would reject an object without metadata wouldn't it? So we don't need to be explicit about it?

Reply:
> I think the API server would reject an object without metadata wouldn't it? So we don't need to be explicit about it?

It should, but the schema is technically incorrect

Not going to mark TypeMeta as controller-gen then adds "" to the lis tof required fields

Reply:
Looks like openapi-gen uses that information for `zz_generated.openapi.go`